lms/djangoapps/certificates/badge_handler.py

```diff
@@ -0,0 +1,134 @@
+    badging service, the other methods don't need to be reproduced.
+    """
+    # Global caching dict
+    badges = {}
+
+    def __init__(self, course_key):
+        self.course_key = course_key
+        assert settings.BADGR_API_TOKEN
+
+    @lazy
+    def base_url(self):
+        """
+        Base URL for all API requests.
+        """
+        return "{}/v1/issuer/issuers/{}".format(settings.BADGR_BASE_URL, settings.BADGR_ISSUER_SLUG)
+
+    @lazy
+    def badge_create_url(self):
+        """
+        URL for generating a new Badge specification
+        """
+        return "{}/badges".format(self.base_url)
+
+    def badge_url(self, mode):
+        """
+        Get the URL for a course's badge in a given mode.
+        """
+        return "{}/{}".format(self.badge_create_url, self.course_slug(mode))
+
+    def assertion_url(self, mode):
+        """
+        URL for generating a new assertion.
+        """
+        return "{}/assertions".format(self.badge_url(mode))
+
+    def course_slug(self, mode):
+        """
+        Slug ought to be deterministic and limited in size so it's not too big for Badgr.
+        """
+        return hashlib.sha256(u"{}{}".format(unicode(self.course_key), unicode(mode))).hexdigest()
+
+    def log_if_raised(self, response):
+        """
+        Log server response if there was an error.
+        """
+        try:
+            response.raise_for_status()
+        except HTTPError:
+            LOGGER.error(
+                u"Encountered an error when contacting the Badgr-Server. Response status was %s.\n%s",
+                response.status_code, response.body
+            )
+            raise
+
+    def get_headers(self):
+        """
+        Headers to send along with the request-- used for authentication.
+        """
+        return {'Authorization': 'Token {}'.format(settings.BADGR_API_TOKEN)}
+
+    def ensure_badge_created(self, mode):
+        """
+        Verify a badge has been created for this mode of the course, and, if not, create it
+        """
+        if self.course_slug(mode) in BadgeHandler.badges:
+            return
+        response = requests.get(self.badge_url(mode), headers=self.get_headers())
+        if response.status_code != 200:
+            self.create_badge(mode)
+        BadgeHandler.badges[self.course_slug(mode)] = True
+
+    def create_badge(self, mode):
+        """
+        Create the badge spec for a course's mode.
+        """
+        course = modulestore().get_course(self.course_key)
+        if mode not in settings.BADGR_IMAGE_SOURCES:
+            raise KeyError("No image specified for course mode {} in BADGR_IMAGE_SOURCES!".format(mode))
+        image_name = settings.BADGR_IMAGE_SOURCES[mode]
+        image = open(image_name, 'rb')
+        # This will fail if the filename is something silly.
+        content_type = 'image/{}'.format(image_name.rsplit('.', 1)[-1].lower())
```

Comment:
@smarnach Could you give me an example of a string that would throw an exception here? I thought I had constructed this to where it wouldn't.

I mean, it may end up constructing a completely nonsense content-type, but it should still upload the file. At which point we'll get an exception either way when the target server complains.